lib/foreman/renderer/scope/macros/loaders.rb

```diff
@@ -0,0 +1,52 @@
+module Foreman
+  module Renderer
+    module Scope
+      module Macros
+        module Loaders
+          include Foreman::Renderer::Errors
+
+          LOADERS = [
+            [ :load_organizations, Organization, :view_organizations ],
```

Comment:
I see a lot of duplicates there... What about make it more declarative? There are many loaders that has only plurarized name of loader and/or name of permission. So what I suggesting is make it more flexible for laziness of developer: 

`LOADERS = [
  Organization, 
  Location,
  Host,
  SmartProxy: {permission: view_smart_proxies, loader: load_smart_proxies}
]` 

and when you're defining method, add some logic to derive names (if it's not hash) with something like:
`permission = "view_#{name.downcase.pluralize}"`

Reply:
I've tried this, but there's no good way of reliable mapping, e.g. load_operating_systems needs Operatingsystem (no camelcase in 2 words) and view_operatingsystems. Also Model -> permission mapping does an SQL query, which is not optimal. Given there's not reliable mapping, we'd end up with exceptions for some resources.

Reply:
I think that's not a problem because It's a great example for apply of Convention over Configuration. If I take this principle to this problem I've got simpler definitions where the exceptions are only detailed configuration. 

The conventions is take the model (ie. `Host`) and try to guess the name of loader (ie. load_hosts) and the name of permission (ie. `:view_hosts`). But there can be different name of loaders (like the collision with existing) and permission can be different too.

So... How to make in the code? We have a loaders. Loaders has permission, model and name. In most cases permission and name is derived from model. That's a some kind of fallback. 
And they are some models with differences like `SmartProxy`. But that's not a problem. You can just define a hash where will be names instead of passing only a model. It's the same with `Operatingsystem`, you just define a hash but only with the name of loader. The other one is just convention. 

And then when you're defining loaders, for hash you take the existing names from that hash and for other cases you just use the convention,

Reply:
So I tried a bit more, but I still think it's not a good improvement

```ruby
LOADERS = [
            { :class => Organization },
            { :class => Location },
            { :class => Host },
            { :class => Operatingsystem, :permission => 'view_operatingsystems', :loader => :load_operating_systems },
            { :class => Subnet },
            { :class => SmartProxy },
            { :class => Usergroup, :permission => 'view_usergroups', :loader => :load_user_groups },
            { :class => Hostgroup, :permission => 'view_hostgroups', :loader => :load_host_groups },
            { :class => Domain },
            { :class => Realm },
            { :class => User },
          ]

          LOADERS.each do |definition|
            resource_name = definition[:class].to_s.downcase.pluralize
            definition[:permission] ||= "view_#{resource_name}"
            definition[:loader] ||= :"load_#{resource_name}"

            define_method definition[:loader] do |search: '', includes: nil, preload: nil, joins: nil, select: nil, batch: 1_000, limit: nil|
              load_resource(klass: definition[:class], search: search, permission: definition[:permission], includes: includes, preload: preload, joins: joins, select: select, batch: batch, limit: limit)
            end
          end
```

this is perhaps acceptable, but then I realized when I needed to allow helpers in lib/foreman/renderer/configuration.rb I need loader names again. I'd either need to duplicate the logic there or not use simple hashes but rather objects representing the tripple and able to calculate missing information. But IMHO brings unnecessary complexity. Given this duplication or fact, we can derive the loader and permission name for most resources is not spanned accross multiple files, I'd prefer to keep as is, and repeat the whole definition for each resources, if you don't insist.

Reply:
What about using a concerns? But understand, it could be using a cannon to shoot a bird so it will be better to stick at the simpler definitions.

Reply:
not sure what you meant with concerns, but yeah, I'd prefer to keep it simple